builtin/audit/stackdriver/backend.go

```diff
@@ -0,0 +1,211 @@
+	// at the appropriate level in the resource heirarchy (project, folder, etc.)
+	parent, ok := conf.Config["parent"]
+	if !ok {
+		return nil, fmt.Errorf("parent is required")
+	}
+
+	logID, ok := conf.Config["log_id"]
+	if !ok {
+		return nil, fmt.Errorf("log_id is required")
+	}
+
+	// Support writing logs asynchronously. This is less secure since there is no guarantee that
+	// the log is written before responding to a request.
+	async := false
+	if asyncRaw, ok := conf.Config["async"]; ok {
+		value, err := strconv.ParseBool(asyncRaw)
+		if err != nil {
+			return nil, err
+		}
+		async = value
+	}
+
+	client, err := logging.NewClient(ctx, parent, option.WithUserAgent(useragent.String()))
+	if err != nil {
+		return nil, err
+	}
+	lg := client.Logger(logID)
+
+	// perform a test log
+	err = lg.LogSync(ctx, logging.Entry{
+		Resource: &mrpb.MonitoredResource{
+			Type: "generic_task",
+		},
+	})
+	if err != nil {
+		return nil, errwrap.Wrapf("Unable to log to stackdriver: {{err}}", err)
+	}
+
+	b := &Backend{
+		client:     client,
+		lg:         lg,
+		async:      async,
+		saltConfig: conf.SaltConfig,
+		saltView:   conf.SaltView,
+		salt:       new(atomic.Value),
+		formatConfig: audit.FormatterConfig{
+			Raw:          logRaw,
+			HMACAccessor: hmacAccessor,
+		},
+	}
+
+	// Ensure we are working with the right type by explicitly storing a nil of
+	// the right type
+	b.salt.Store((*salt.Salt)(nil))
+
+	b.formatter.AuditFormatWriter = &audit.JSONFormatWriter{
+		SaltFunc: b.Salt,
+	}
+
+	return b, nil
+}
+
+// Backend is the audit backend for Stackdriver audit logging.
+type Backend struct {
+	client *logging.Client
+	lg     *logging.Logger
+	async  bool
+
+	formatter    audit.AuditFormatter
+	formatConfig audit.FormatterConfig
+
+	saltMutex  sync.RWMutex
+	salt       *atomic.Value
+	saltConfig *salt.Config
+	saltView   logical.Storage
+}
+
+var _ audit.Backend = (*Backend)(nil)
+
+func (b *Backend) Salt(ctx context.Context) (*salt.Salt, error) {
+	s := b.salt.Load().(*salt.Salt)
+	if s != nil {
+		return s, nil
+	}
+
+	b.saltMutex.Lock()
+	defer b.saltMutex.Unlock()
+
+	s = b.salt.Load().(*salt.Salt)
+	if s != nil {
+		return s, nil
+	}
+
+	newSalt, err := salt.NewSalt(ctx, b.saltView, b.saltConfig)
+	if err != nil {
+		b.salt.Store((*salt.Salt)(nil))
+		return nil, err
+	}
+
+	b.salt.Store(newSalt)
+	return newSalt, nil
+}
+
+func (b *Backend) GetHash(ctx context.Context, data string) (string, error) {
+	salt, err := b.Salt(ctx)
+	if err != nil {
+		return "", err
+	}
+	return audit.HashString(salt, data), nil
+}
+
+func (b *Backend) LogRequest(ctx context.Context, in *logical.LogInput) error {
+	var buf bytes.Buffer
+	b.formatter.FormatRequest(ctx, &buf, b.formatConfig, in)
+	log := logging.Entry{
+		Payload: json.RawMessage(buf.String()),
+		Resource: &mrpb.MonitoredResource{
+			Type: "generic_task",
+		},
+	}
+	if b.async {
+		b.lg.Log(log)
+		return nil
+	}
+	return b.lg.LogSync(ctx, log)
```

Comment:
According to the [godocs](https://pkg.go.dev/cloud.google.com/go/logging?tab=doc#Logger.LogSync) `LogSync` blocks and can be slow and thus [not recommended](https://pkg.go.dev/cloud.google.com/go/logging?tab=doc#hdr-Synchronous_Logging) for normal usage, should this be a concern? I am not sure exactly how slow this translates to under normal operation (e.g. milliseconds vs seconds).